pkg/controller/encryption/ipsec/ipsec_controller.go

```diff
@@ -0,0 +1,526 @@
+	}
+	for _, addr := range localNode.Status.Addresses {
+		if strings.Compare(string(addr.Type), string(v1.NodeInternalIP)) == 0 {
+			ipsecController.kmeshNodeInfo.Spec.Addresses = append(ipsecController.kmeshNodeInfo.Spec.Addresses, addr.Address)
+			ipsecController.ipsecHandler.SetNodeInfo(addr.Address, ipsecController.kmeshNodeInfo.Spec.BootID, ipsecController.ipsecHandler.Spi, true, localNode.Spec.PodCIDRs)
+		}
+	}
+
+	if _, err := nodeinfoInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ipsecController.handleKNIAdd(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			ipsecController.handleKNIUpdate(oldObj, newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			ipsecController.handleKNIDelete(obj)
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add event handler to kmeshnodeinfoInformer: %v", err)
+	}
+
+	return ipsecController, nil
+}
+
+func (c *IPSecController) Run(stop <-chan struct{}) {
+	defer c.queue.ShutDown()
+	go c.informer.Run(stop)
+	if !cache.WaitForCacheSync(stop, c.informer.HasSynced) {
+		log.Error("timed out waiting for caches to sync")
+		return
+	}
+
+	if err := c.attachTcDecrypt(); err != nil {
+		log.Errorf("%v", err)
+		return
+	}
+
+	// create xfrm in rule, current host not update my kmeshnodeinfo
+	// the peer end does not use the key of the current host to send encrypted data.
+	if err := c.syncAllNodeInfo(); err != nil {
+		log.Errorf("failed to sync all node info: %v", err)
+		return
+	}
+
+	// update my kmesh node info, notify other machines that the key can be updated.
+	if err := c.updateLocalKmeshNodeInfo(); err != nil {
+		log.Errorf("failed to update local node info: %v", err)
+		return
+	}
+
+	c.ipsecHandler.StartWatch(c.handleIpsecUpdate)
+
+	go wait.Until(func() {
+		for c.processNextItem() {
+		}
+	}, 0, stop)
+
+	<-stop
+}
+
+func (c *IPSecController) Stop() {
+	c.ipsecHandler.StopWatch()
+	if restart.GetStartType() == restart.Normal {
+		c.knclient.Delete(context.TODO(), c.kmeshNodeInfo.Name, metav1.DeleteOptions{})
+		c.detachTcDecrypt()
+		c.CleanAllIPsec()
+	}
+}
+
+func (c *IPSecController) handleTc(mode int) error {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return fmt.Errorf("failed to get interfaces: %v", err)
+	}
+	for _, iface := range ifaces {
+		if iface.Flags&net.FlagLoopback != 0 {
+			continue
+		}
+
+		addresses, err := iface.Addrs()
+		if err != nil {
+			log.Warnf("failed to get interface %v address: %v", iface.Name, err)
+			continue
+		}
+
+		link, err := netlink.LinkByName(iface.Name)
+		if err != nil {
+			log.Warnf("failed to link interface %v, %v", iface, err)
+			continue
+		}
+
+		for _, rawAddr := range addresses {
+			addr, ok := rawAddr.(*net.IPNet)
+			if !ok {
+				log.Warnf("failed to convert ifaddr %v, %v", rawAddr, err)
+				continue
+			}
+			for _, rawLocalAddr := range c.kmeshNodeInfo.Spec.Addresses {
+				localAddr := net.ParseIP(rawLocalAddr)
+				if addr.IP.Equal(localAddr) {
```

Comment:
what does this check mean

Reply:
Mount the tc program only when NodeInternalIP exists on the NIC.